ksqldb-common/src/main/java/io/confluent/ksql/schema/ksql/SystemColumns.java

```diff
@@ -15,39 +15,78 @@
 
 package io.confluent.ksql.schema.ksql;
 
+import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
 import io.confluent.ksql.name.ColumnName;
 import io.confluent.ksql.schema.ksql.types.SqlType;
 import io.confluent.ksql.schema.ksql.types.SqlTypes;
+import io.confluent.ksql.util.KsqlException;
+import java.util.HashSet;
+import java.util.Map;
 import java.util.Set;
 
 public final class SystemColumns {
 
   public static final ColumnName ROWKEY_NAME = ColumnName.of("ROWKEY");
-
   public static final ColumnName ROWTIME_NAME = ColumnName.of("ROWTIME");
   public static final SqlType ROWTIME_TYPE = SqlTypes.BIGINT;
 
+  public static final ColumnName ROWOFFSET_NAME = ColumnName.of("ROWOFFSET");
+  public static final SqlType ROWOFFSET_TYPE = SqlTypes.BIGINT;
+
+  public static final ColumnName ROWPARTITION_NAME = ColumnName.of("ROWPARTITION");
+  public static final SqlType ROWPARTITION_TYPE = SqlTypes.INTEGER;
+
   public static final ColumnName WINDOWSTART_NAME = ColumnName.of("WINDOWSTART");
   public static final ColumnName WINDOWEND_NAME = ColumnName.of("WINDOWEND");
   public static final SqlType WINDOWBOUND_TYPE = SqlTypes.BIGINT;
 
-  public static final int LEGACY_PSEUDOCOLUMN_VERSION_NUMBER = 0;
-  public static final int CURRENT_PSEUDOCOLUMN_VERSION_NUMBER = 5;
 
-  private static final Set<ColumnName> PSEUDO_COLUMN_NAMES = ImmutableSet.of(
-      ROWTIME_NAME
-  );
+  public static final int ROWTIME_PSEUDOCOLUMN_VERSION = 0;
+  public static final int ROWPARTITION_ROWOFFSET_PSEUDOCOLUMN_VERSION = 1;
+
+  public static final int LEGACY_PSEUDOCOLUMN_VERSION_NUMBER = ROWTIME_PSEUDOCOLUMN_VERSION;
+  public static final int CURRENT_PSEUDOCOLUMN_VERSION_NUMBER =
+      ROWPARTITION_ROWOFFSET_PSEUDOCOLUMN_VERSION;
+
+  public static Set<ColumnName> getPseudoColumnsFromVersion(final int pseudoColumnVersion) {
+    if (!VERSIONED_PSEUDO_COLUMNS.containsKey(pseudoColumnVersion)) {
+      throw new KsqlException("Provided pseudoColumnVersion has no corresponding columns defined");
+    }
+    return VERSIONED_PSEUDO_COLUMNS.get(pseudoColumnVersion);
+  }
+
+  private static final Set<ColumnName> VERSION_ZERO_NAMES;
+  private static final Set<ColumnName> VERSION_ONE_NAMES;
+
+  static {
+    VERSION_ZERO_NAMES = new HashSet<>();
+    VERSION_ZERO_NAMES.add(ROWTIME_NAME);
+
+    VERSION_ONE_NAMES = new HashSet<>(VERSION_ZERO_NAMES);
+    VERSION_ONE_NAMES.add(ROWPARTITION_NAME);
+    VERSION_ONE_NAMES.add(ROWOFFSET_NAME);
+  }
+
+  private static final Map<Integer, Set<ColumnName>> VERSIONED_PSEUDO_COLUMNS = ImmutableMap.of(
```

Comment:
nit: rename to `PSEUDO_COLUMN_NAMES_BY_VERSION` to be consistent with `SYSTEM_COLUMN_NAMES_BY_VERSION` below.